ssl/ssl_test.cc

```diff
@@ -8853,6 +8853,353 @@ TEST(SSLTest, ProcessTLS13NewSessionTicket) {
+// processed in server tls state machine, |PHA_Config| is initialized,
+// and sigalgs + client CA data is copied appropriately to |PHA_Config|
+TEST(SSLTest, ImmediatePHA) {
+  // Configure client and server to negotiate TLS 1.3 only.
+  bssl::UniquePtr<SSL_CTX> client_ctx(CreateContextWithTestCertificate(TLS_method()));
+  bssl::UniquePtr<SSL_CTX> server_ctx(
+      CreateContextWithTestCertificate(TLS_method()));
+  ASSERT_TRUE(client_ctx);
+  ASSERT_TRUE(server_ctx);
+  ASSERT_TRUE(SSL_CTX_set_min_proto_version(client_ctx.get(), TLS1_3_VERSION));
+  ASSERT_TRUE(SSL_CTX_set_min_proto_version(server_ctx.get(), TLS1_3_VERSION));
+  ASSERT_TRUE(SSL_CTX_set_max_proto_version(client_ctx.get(), TLS1_3_VERSION));
+  ASSERT_TRUE(SSL_CTX_set_max_proto_version(server_ctx.get(), TLS1_3_VERSION));
+
+  // Configure server to ask for client authentication immediately after the
+  // handshake
+  SSL_CTX_set_verify(server_ctx.get(), SSL_VERIFY_PEER | SSL_VERIFY_POST_HANDSHAKE, nullptr);
+
+  // Configure trust store for server.
+  SSL_CTX_set_client_CA_list(server_ctx.get(), get_test_cert_store());
+
+  bssl::UniquePtr<SSL> client, server;
+  ASSERT_TRUE(CreateClientAndServer(&client, &server, client_ctx.get(),
+                                    server_ctx.get()));
+
+  // Enable sending pha_ext on client side
+  SSL_set_post_handshake_auth(client.get(), 1);
+
+  // Start the handshake, client first flight and write. Sends the extension
+  EXPECT_TRUE(client.get()->s3->pha_enabled);
+  EXPECT_TRUE(client.get()->s3->pha_ext == SSL_PHA_NONE);
+  SSL_do_handshake(client.get());
+
+  EXPECT_TRUE(client.get()->s3->pha_ext == SSL_PHA_EXT_SENT);
+  EXPECT_TRUE(server.get()->s3->pha_ext == SSL_PHA_NONE);
+
+  // First server flight, process client hello and send Server Hello,
+  // Certificate, CertificateVerify and configure immediate PHA
+  SSL_do_handshake(server.get());
+  // Since |SSL_VERIFY_POST_HANDSHAKE| is set, request pending
+  EXPECT_TRUE(server.get()->s3->pha_ext == SSL_PHA_REQUEST_PENDING);
+  // should be false so CertificateRequest not sent in initial message
+  EXPECT_FALSE(server.get()->s3->hs->cert_request);
+
+  // Second client flight, process server messages and send client Finished
+  SSL_do_handshake(client.get());
+  EXPECT_TRUE(client.get()->s3->pha_ext == SSL_PHA_EXT_SENT);
+
+  // Storing sigalgs since handshake object is destroyed in next server processing
+  Span<const uint16_t> verify_sigalgs = tls13_get_verify_sigalgs_pha(server.get()->s3->hs.get());
+  // Storing clientCA list before copy
+  const STACK_OF(CRYPTO_BUFFER) *names = ssl_get_client_CAs_pha(server.get()->s3->hs.get());
+
+  // Second server flight, process client Finished and put CertificateRequest
+  // on the wire to be flushed on the next Server write
+  SSL_do_handshake(server.get());
+  EXPECT_TRUE(server.get()->s3->pha_ext == SSL_PHA_REQUESTED);
+
+  // Check initialization of PHA_config struct
+  EXPECT_TRUE(server.get()->s3->pha_config != nullptr);
```

Comment:
```suggestion
  EXPECT_TRUE(server.get()->s3->pha_config);
```

* We can also use `EXPECT_EQ` instead in some places in this test file